server/src/compression.rs

```diff
@@ -79,6 +85,79 @@ pub fn try_compress(
     Ok(compressed_record_batch_builder.finish())
 }
 
+/// Merges the segments in `compressed_segments` that contain equivalent models.
+/// Assumes that the segments in `compressed_segments` are all from the same
+/// time series and that the segments are all sorted according to time.
+pub fn merge_segments(compressed_segments: RecordBatch) -> RecordBatch {
+    // TODO: merge segments with none equivalent models.
+
+    // Extract the columns from the RecordBatch.
+    let model_type_ids = crate::get_array!(compressed_segments, 0, UInt8Array);
+    let timestamps = crate::get_array!(compressed_segments, 1, BinaryArray);
+    let start_times = crate::get_array!(compressed_segments, 2, TimestampArray);
+    let end_times = crate::get_array!(compressed_segments, 3, TimestampArray);
+    let values = crate::get_array!(compressed_segments, 4, BinaryArray);
+    let min_values = crate::get_array!(compressed_segments, 5, ValueArray);
+    let max_values = crate::get_array!(compressed_segments, 6, ValueArray);
+    let errors = crate::get_array!(compressed_segments, 7, Float32Array);
+
+    // For each segment, check if it can be merged with another segment.
+    let num_rows = compressed_segments.num_rows();
+    let mut compressed_segments_to_merge = HashMap::with_capacity(num_rows);
+
+    for index in 0..compressed_segments.num_rows() {
+        // f32 are converted to u32 with the same bitwise representation as f32
+        // and f64 does not implement std::hash::Hash and thus cannot be hashed.
+        let model = (
+            model_type_ids.value(index),
+            values.value(index),
+            min_values.value(index).to_bits(),
+            max_values.value(index).to_bits(),
+        );
+
+        compressed_segments_to_merge
+            .entry(model)
+            .or_insert_with(|| Vec::new())
+            .push(index);
+    }
+
+    // If none of the segments can be merged return the original compressed
+    // segments, otherwise return the smaller set of merged compressed segments.
+    if compressed_segments_to_merge.len() < num_rows {
+        let mut merged_compressed_segments = CompressedSegmentBatchBuilder::new(num_rows);
+        for (_, indexes) in compressed_segments_to_merge {
+            // Merge timestamps.
+            let mut timestamp_arrays = Vec::with_capacity(indexes.len());
+            for index in &indexes {
+                let start_time = start_times.value(*index);
+                let end_time = end_times.value(*index);
+                let timestamps = timestamps.value(*index);
+                timestamp_arrays.push(timestamps::decompress_all_timestamps(
+                    start_time, end_time, timestamps,
+                ));
+            }
+            let timestamps = flatten_timestamp_arrays(timestamp_arrays);
+            let compressed_timestamps = timestamps::compress_residual_timestamps(&timestamps);
```

Comment:
If two regular but non-sequential segments are merged would this then turn the merged segment irregular? and if so does this mean that merging could in some cases increase the size of the segment?

Reply:
Yes and yes until we extend the compression used for timestamps from regular time series to support timestamps from regular time series with gaps. The main problem to solve is how to efficiently encode `one-bit flag, n-bit length, n-bit gap, n-bit length, etc` using the smallest number of bits while still using an encoding that is somewhat efficient to decode. With this extension it should always be beneficial to merge segments as the amount of storage required to store the lengths and gaps should be lower than the amount of storage required for another segment's metadata and model.

Reply:
Makes sense.